### PR TITLE
Add Session-based Recommendations With Recurrent Neural Networks (GRU4Rec) model

### PR DESCRIPTION
**Pull request by lthoang**
### Description

This is an adaptation of an official implementation of [GRU4Rec](https://github.com/hidasib/GRU4Rec_PyTorch_Official).

### Checklist:

- [ ] I have added tests.
- [x] I have updated the documentation accordingly.
- [x] I have updated `README.md` (if you are adding a new model).
- [x] I have updated `examples/README.md` (if you are adding a new example).
- [ ] I have updated `datasets/README.md` (if you are adding a new dataset).

### CODE REVIEW

File: cornac/models/gru4rec/gru4rec.py [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+        if training:
+            E = self.DE(E)
+        if not (self.constrained_embedding or self.embedding):
+            H[0] = E
+        Xh = self.hidden_step(E, H, training=training)
+        R = self.score_items(Xh, O, B)
+        return R
+
+
+def io_iter(
```
Comment (tqtg) [inline cornac/models/gru4rec/gru4rec.py:310]: Is it something that can be reused for other models?

File: cornac/models/gru4rec/recom_gru4rec.py [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+                    H[i][np.nonzero(start_mask)[0], :] = 0
+                    H[i].detach_()
+                    H[i] = H[i][valid_id]
+                in_iids = torch.tensor(in_iids, requires_grad=False, device=self.device)
+                out_iids = torch.tensor(
+                    out_iids, requires_grad=False, device=self.device
+                )
+                self.model.zero_grad()
+                R = self.model.forward(in_iids, H, out_iids, training=True)
+                L = self.loss_function(R, out_iids, len(in_iids)) / self.batch_size
```
Comment (tqtg) [inline cornac/models/gru4rec/recom_gru4rec.py:265]: I would suggest to move all functions related to defining the loss criteria to the `gru4rec.py`. That would improve code structure and readability quite a bit.

File: cornac/models/spop/recom_spop.py [hunk 3]
```diff
@@ -48,7 +48,7 @@ def fit(self, train_set, val_set=None):
         return self
 
     def score(self, user_idx, history_items, **kwargs):
-        item_scores = np.ones(self.total_items, dtype=np.float32)
+        item_scores = np.zeros(self.total_items, dtype=np.float32)
```
Comment (tqtg) [inline cornac/models/spop/recom_spop.py:51]: What is the cause for this change? Shall we create a separate PR for this instead?

**Comment by tqtg** [comment]
@lthoang please also consider adding model into the [docs](https://github.com/PreferredAI/cornac/blob/master/docs/source/api_ref/models.rst)

File: cornac/models/gru4rec/gru4rec.py (hunk 1 above)
Comment (lthoang) [reply cornac/models/gru4rec/gru4rec.py:310]: Currently, there is no other model reuse this iteration. We may need to migrate it to `dataset` when there is another model reuse this iteration.

File: cornac/models/gru4rec/recom_gru4rec.py (hunk 2 above)
Comment (lthoang) [reply cornac/models/gru4rec/recom_gru4rec.py]: Updated.

File: cornac/models/spop/recom_spop.py (hunk 3 above)
Comment (lthoang) [reply cornac/models/spop/recom_spop.py]: I have created another PR for this fix.

**Review by tqtg — APPROVED** [review]

**Comment by tqtg** [comment]
LGTM

**Comment by tqtg** [comment]
@lthoang @hieuddo just curious whether we obtain comparable results to the original implementation on the same dataset?

**Comment by lthoang** [comment]
> @lthoang @hieuddo just curious whether we obtain comparable results to the original implementation on the same dataset?

To be able to compare with original implementation, we need to further add an evaluation option for evaluating every next item in test sequence.
I have mentioned in https://github.com/PreferredAI/cornac/issues/543#issuecomment-1860831122

@tqtg Should we also include that in this PR, or create a separate PR?

**Comment by tqtg** [comment]
@lthoang let have this merged and we can work on another PR for that evaluation feature
